osquery/tables/system/windows/windows_optional_features.cpp

```diff
@@ -0,0 +1,85 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#include <string>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/regex.hpp>
+
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include "osquery/core/conversions.h"
+#include "osquery/core/windows/wmi.h"
+
+namespace osquery {
+namespace tables {
+
+std::string getDismPackageFeatureStateName(uint32_t state);
+
+/**
+ * return collection of Windows Features installation states.
+ * On Window 10 Pro, returns about 140 rows.
+ */
+QueryData genWinOptionalFeatures(QueryContext& context) {
+  QueryData results;
+
+  const WmiRequest wmiReq(
+      "SELECT Caption,Name,InstallState FROM Win32_OptionalFeature");
+  const std::vector<WmiResultItem>& wmiResults = wmiReq.results();
+
+  if (wmiResults.empty()) {
+    return results;
+  }
+
+  for (unsigned int i = 0; i < wmiResults.size(); ++i) {
+    Row r;
+    uint32_t state = 99;
+
+    wmiResults[i].GetString("Name", r["name"]);
+    wmiResults[i].GetString("Caption", r["caption"]);
+
+    // wbemtest.exe shows Column as UINT32, but comes in as I4.
+    // For whatever reason, I4 is accessed using GetLong().
+
+    if (wmiResults[i].GetUnsignedInt32("InstallState", state).ok() == false) {
+      long i4;
+      if (wmiResults[i].GetLong("InstallState", i4).ok()) {
+        state = (uint32_t)i4;
+      }
+    }
+    r["state"] = INTEGER(state);
+
+    r["statename"] = getDismPackageFeatureStateName(atoi(r["state"].c_str()));
```

Comment:
Hrm, the extra function, as well as the tests for it, seem a bit unnecessary to me.

Why not leverage a gloabl static `const std::map<unsigned int, std::string>`? You could grab the `state` value, regardless of returned type, coerce it to an `unsigned int/long`, then perform a `std::map` lookup on the `.keys()` and if your desired keys not present return `unknown`? This will save you some complexity and stack allocations. As for adding in tests you could leverage the [integration sanity tests](https://github.com/osquery/osquery/blob/master/tests/integration/tables/wmi_bios_info.cpp) to ensure that the data is at least returning as you expect it to, what do you think?

Reply:
yeah, I just wanted to keep the flow clean here and be able to unit test the name lookup.